File: app/src/main/kotlin/tools/isekai/terminal/SnippetCommands.kt
package tools.isekai.terminal

import tools.isekai.terminal.data.Snippet

/**
 * スニペット／接続後自動実行コマンドの文字列をターミナルへ送信するバイト列に変換する純粋関数群。
 * Android/Rust いずれにも依存しないため単体テストが容易。
 */
object SnippetCommands {
    /**
     * [command] の各行区切り (`\n`, `\r\n`) を `\r` (CR) に正規化して UTF-8 バイト列にする。
     * [appendNewline] が true かつ末尾がまだ `\r` で終わっていなければ、末尾にも `\r` を追加する
     * （＝最後の行も Enter される）。false なら最後の行は Enter されずに残る。
     */
    fun toBytes(command: String, appendNewline: Boolean = true): ByteArray {
        if (command.isEmpty()) return ByteArray(0)
        val normalized = command.replace("\r\n", "\n").replace('\n', '\r')
        val withTrailing =
            if (appendNewline && !normalized.endsWith('\r')) normalized + '\r' else normalized
        return withTrailing.toByteArray(Charsets.UTF_8)
    }

    fun toBytes(snippet: Snippet): ByteArray = toBytes(snippet.command, snippet.appendNewline)
}

File: ios/Sources/IsekaiTerminalCoreLogic/SnippetCommands.swift
import Foundation

/// Phase 1G-1(#53): スニペットのコマンド文字列をターミナルへ送信するバイト列に変換する
/// 純粋関数。Android版`SnippetCommands.toBytes`と対称(Rust/UIに依存しないため単体テストが容易)。
/// GRDBの`Snippet`レコード型に依存するオーバーロードは`IsekaiTerminalCore`側
/// (`Sources/IsekaiTerminalCore/ProfileDatabase.swift`)の`extension`で追加している。
public enum SnippetCommands {
    /// `command`の各行区切り(`\n`, `\r\n`)を`\r`(CR)に正規化してUTF-8バイト列にする。
    /// `appendNewline`がtrueかつ末尾がまだ`\r`で終わっていなければ、末尾にも`\r`を追加する
    /// (=最後の行もEnterされる)。falseなら最後の行はEnterされずに残る。
    public static func toBytes(command: String, appendNewline: Bool = true) -> Data {
        guard !command.isEmpty else { return Data() }
        let normalized = command.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\n", with: "\r")
        let withTrailing = (appendNewline && !normalized.hasSuffix("\r")) ? normalized + "\r" : normalized
        return Data(withTrailing.utf8)
    }
}
